pkg/model/utxo/dust_db.go

```diff
@@ -0,0 +1,223 @@
+		switch output.outputType {
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
+			diff, found := dustDiff[address]
+			if found {
+				diff.DustAllowanceBalanceDiff += int64(output.Amount())
+			} else {
+				dustDiff[address] = NewDustDiff(int64(output.Amount()), 0)
+			}
+		case iotago.OutputSigLockedSingleOutput:
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
+				diff, found := dustDiff[address]
+				if found {
+					diff.DustOutputCount += 1
+				} else {
+					dustDiff[address] = NewDustDiff(0, 1)
+				}
+			}
+		}
+	}
+
+	for _, spent := range newSpents {
+
+		// Remove spent dust
+		output := spent.Output()
+		switch output.outputType {
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
+			diff, found := dustDiff[address]
+			if found {
+				diff.DustAllowanceBalanceDiff -= int64(output.Amount())
+			} else {
+				dustDiff[address] = NewDustDiff(-int64(output.Amount()), 0)
+			}
+		case iotago.OutputSigLockedSingleOutput:
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
+				diff, found := dustDiff[address]
+				if found {
+					diff.DustOutputCount -= 1
+				} else {
+					dustDiff[address] = NewDustDiff(0, -1)
+				}
+			}
+		}
+	}
+
+	return u.applyDustDiff(dustDiff)
+}
+
+func (u *Manager) rollbackDustWithoutLocking(newOutputs Outputs, newSpents Spents) error {
+
+	dustDiff := make(map[string]*DustDiff)
+
+	// we have to delete the newOutputs of this milestone
+	for _, output := range newOutputs {
+
+		// Remove unspent dust
+		switch output.outputType {
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
+			diff, found := dustDiff[address]
+			if found {
+				diff.DustAllowanceBalanceDiff -= int64(output.Amount())
+			} else {
+				dustDiff[address] = NewDustDiff(-int64(output.Amount()), 0)
+			}
+		case iotago.OutputSigLockedSingleOutput:
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
+				diff, found := dustDiff[address]
+				if found {
+					diff.DustOutputCount -= 1
+				} else {
+					dustDiff[address] = NewDustDiff(0, -1)
+				}
+			}
+		}
+	}
+
+	// we have to store the spents as output and mark them as unspent
+	for _, spent := range newSpents {
+
+		// Re-Add previously-spent dust
+		output := spent.Output()
+		switch output.outputType {
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
+			diff, found := dustDiff[address]
+			if found {
+				diff.DustAllowanceBalanceDiff += int64(output.Amount())
+			} else {
+				dustDiff[address] = NewDustDiff(int64(output.Amount()), 0)
+			}
+		case iotago.OutputSigLockedSingleOutput:
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
```

Comment:
you should convert the address bytes to string, but the String function uses hex encoding